Make SemanticVersion compatible with SemVer 2.0

iSazonov opened the pull request:
Fix #5004 

### Fix description

Made compatible with SymVer 2.0 http://semver.org/ (p.10)
Fixed comparisons
Refactored tests 
Added new tests

### Additional information
Useful sample https://github.com/maxhauser/semver/blob/master/src/Semver/SemVersion.cs
Related #2983

iSazonov committed:
Make SemanticVersion compatible with SymVer 2.0

daxian-dbw commented:
@JamesWTruher @PaulHigin Can you please review this PR?

PaulHigin commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -571,77 +659,127 @@ public static bool TryParse(string version, out SemanticVersion result)
+            string buildLabel = null;
+
             var dashIndex = version.IndexOf('-');
+            var plusIndex = version.IndexOf('+');
```
This parse code would be easier to read if it included comments with examples of expected version string.

iSazonov replied:
Comment added.

PaulHigin commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -571,77 +659,127 @@ public static bool TryParse(string version, out SemanticVersion result)
+                }
+                else
+                {
+                    // No preLabel - preLabel == null;
```
Can you add a comment that a hyphen character ('-') is allowed as part of the buildLabel (metadata) per the spec?  BTW why do you use "buildLabel" instead of "metadata" or "metadataLable" as referred to by the spec?

iSazonov replied:
Comments added and edited.

PaulHigin commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -571,77 +659,127 @@ public static bool TryParse(string version, out SemanticVersion result)
             }
 
-            if (parsedComponent < 0)
+            if (preLabel != null && !Regex.IsMatch(preLabel, s_LabelUnitRegEx) ||
```
I wonder if we should do this preLabel, buildLabel regex validation in the SemanticVersion constructor.  That way we can prevent malformed SemanticVersion objects from being created.

iSazonov replied:
I wonder too why I checked only first char in constructors :-) I wonted to be fast. We should use RegEx.

Fixed.

PaulHigin commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -748,7 +882,6 @@ public override int GetHashCode()
         /// </summary>
         public static bool operator <(SemanticVersion v1, SemanticVersion v2)
         {
-            if ((object)v1 == null) throw PSTraceSource.NewArgumentException(nameof(v1));
```
I feel we still need this check, otherwise v1.CompareTo will throw null reference exception.

iSazonov replied:
Fixed by new method Compare()

PaulHigin commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -757,7 +890,6 @@ public override int GetHashCode()
         /// </summary>
         public static bool operator <=(SemanticVersion v1, SemanticVersion v2)
         {
-            if ((object)v1 == null) throw PSTraceSource.NewArgumentException(nameof(v1));
```
I feel we still need this check, otherwise v1.CompareTo will throw null reference exception.

iSazonov replied:
Fixed.

PaulHigin commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -766,15 +898,50 @@ public override int GetHashCode()
         public static bool operator >(SemanticVersion v1, SemanticVersion v2)
         {
-            return (v2 < v1);
+            return (v1.CompareTo(v2) > 0);
```
We should check v1 for null value.

iSazonov replied:
Fixed.

PaulHigin commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -766,15 +898,50 @@ public override int GetHashCode()
         public static bool operator >=(SemanticVersion v1, SemanticVersion v2)
         {
-            return (v2 <= v1);
+            return (v1.CompareTo(v2) >= 0);
```
We should check v1 for null value.

iSazonov replied:
Fixed.

PaulHigin reviewed (changes requested):
I am not quite finished with the review.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -384,12 +386,18 @@ IEnumerator IEnumerable.GetEnumerator()
     /// </summary>
     public sealed class SemanticVersion : IComparable, IComparable<SemanticVersion>, IEquatable<SemanticVersion>
     {
+        private static readonly string s_VersionSansRegEx = @"^(?<major>\d+)(\.(?<minor>\d+))?(\.(?<patch>\d+))?$";
```
Fixed.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -398,8 +406,47 @@ public SemanticVersion(string version)
+        /// </summary>
+        /// <param name="major">The major version</param>
+        /// <param name="minor">The minor version</param>
+        /// <param name="patch">The minor version</param>
```
Fixed.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -398,8 +406,47 @@ public SemanticVersion(string version)
+            }
+            else
+            {
+                PreLabel = null;
```
I only thought about symmetry and better readability.

Removed.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -410,17 +457,27 @@ public SemanticVersion(string version)
-        /// <exception cref="PSArgumentNullException">
-        /// If <paramref name="label"/> is null or an empty string.
+        /// <exception cref="FormatException">
+        /// If <paramref name="label"/> starts with '-'.
```
Complex RegEx is a puzzle for me too. I found some sample RegEx strings for SymVer but after testing I concluded that they all have bugs. 
So I split the SymVer regex string on some string (s_VersionSansRegEx, s_LabelRegEx, s_LabelUnitRegEx). This slightly increased the code but helped to avoid bugs.

Comment added.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -521,22 +602,29 @@ public SemanticVersion(Version version)
+        ///
         /// </summary>
-        public string Label { get; }
+        //public string Label { get; }
```
Removed.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -521,22 +602,29 @@ public SemanticVersion(Version version)
-        public string Label { get; }
+        //public string Label { get; }
+
+        public string PreLabel { get; }
```
Comments added.

iSazonov committed:
Address feedback

Replace static strings with const.
Add strong checks in constructors.
Add comments.
Add static Compare() and fix comparisions operators.

PaulHigin commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -680,16 +822,14 @@ public int CompareTo(SemanticVersion value)
             if (Patch != value.Patch)
                 return Patch > value.Patch ? 1 : -1;
 
-            if (Label == null)
```
Actually this PreLabel null check here can be removed since it is duplicated in ComparePreLabel().

PaulHigin replied:
But please add the comment in ComparePreLabel() method.

iSazonov replied:
Duplicate checks was removed.
Comment was added.

iSazonov committed:
Remove duplicate checks and add comment.

iSazonov commented:
I've rebased to resolve a merge conflict.

daxian-dbw commented:
@anmenaga Can you please also review this PR? Thanks!

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -571,77 +647,142 @@ public static bool TryParse(string version, out SemanticVersion result)
+        /// </summary>
+        public static int Compare(SemanticVersion versionA, SemanticVersion versionB)
+        {
+            if (versionA == null) return -1;
```
what about versionA and versionB are both null?

iSazonov replied:
Good catch!

Fixed.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -705,21 +839,18 @@ public override bool Equals(object obj)
-                Minor.GetHashCode(),
-                Patch.GetHashCode(),
-                Label == null ? 0 : Label.GetHashCode());
+            return versionString.GetHashCode();
```
`versionString` could be null when running this method.
Besides, if we use the hashcode of  `versionString` for the SemanticVersion, we would have a hashcode collision, would that be a problem? Why not continue to use `Utils.CombineHashCodes`?

iSazonov replied:
null - fixed.

versionString is relatively short string - why you believe that CoreFX implementation of `String.GetHashCode()` so bad that we should make more complex hashcode?

daxian-dbw replied:
What I mean is the hashcode of a SemanticsVersion instance would be the same of the hashcode of a string (the versionString), so there is a collision in terms that 2 different objects have the same hashcode.

iSazonov replied:
In the [docs](https://msdn.microsoft.com/en-us/library/system.object.gethashcode%28v=vs.110%29.aspx) I don't found that the hash must be different for different objects of different types. It seems we follow the docs - it will be work well with Dictionary and Hashtable types.

daxian-dbw replied:
Maybe I'm just overthinking it. Let's keep it as is.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -748,33 +879,67 @@ public override int GetHashCode()
+                }
+                else
+                {
+                    if (isNumber1) { return -1; }
```
It would be great if you can add comment here saying that "Numeric identifiers always have lower precedence than non-numeric identifiers. "

iSazonov replied:
I copy-paste all text from the standard.

iSazonov committed:
Remove "s_' prefix from constant names

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -384,12 +386,18 @@ IEnumerator IEnumerable.GetEnumerator()
     {
+        private const string s_VersionSansRegEx = @"^(?<major>\d+)(\.(?<minor>\d+))?(\.(?<patch>\d+))?$";
+        private const string s_LabelRegEx = @"^((?<preLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?(\+(?<buildLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?$";
+        private const string s_LabelUnitRegEx = @"^[0-9A-Za-z][0-9A-Za-z\-\.]*?$";
```
`s_` was removed.
`?` was removed - I think it is bug in any case.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -398,8 +406,39 @@ public SemanticVersion(string version)
+        /// <param name="preLabel">The preLabel for the version</param>
+        /// <param name="buildLabel">The buildLabel for the version</param>
+        /// <exception cref="FormatException">
+        /// If <paramref name="preLabel"/> starts with '-' ot contains '+'.
```
I changed the comments because the code was changed previously.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -521,22 +591,28 @@ public SemanticVersion(Version version)
-        /// The last component in a SemanticVersion - may be null if not specified.
+        /// PreLabel position in the SymVer string 'major.minor.patch-PreLabel+BuildLabel'.
+        /// </summary>
+
```
Done.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -521,22 +591,28 @@ public SemanticVersion(Version version)
+        /// PreLabel position in the SymVer string 'major.minor.patch-PreLabel+BuildLabel'.
+        /// </summary>
+
+        public string PreLabel { get; }
```
Replaced with `PreReleaseLabel` - it looks more readable.

iSazonov committed:
Fix comments and names.

iSazonov committed:
Fix Compare(), GetHashCode() and add comments.

iSazonov committed:
Fix tests

daxian-dbw commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/DotNetTypes_format_ps1xml.cs:
```diff
@@ -478,6 +486,46 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_System_Version()
+                        .AddPropertyColumn("Major")
+                        .AddPropertyColumn("Minor")
+                        .AddPropertyColumn("Patch")
+                        .AddPropertyColumn("PreLabel")
```
The property name has changed to 'PrereleaseLabel'.

iSazonov replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/DotNetTypes_format_ps1xml.cs:
```diff
@@ -478,6 +486,46 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_System_Version()
+                        .AddPropertyColumn("Minor")
+                        .AddPropertyColumn("Build")
+                        .AddPropertyColumn("Revision")
+                        .AddPropertyColumn("PSSemanticVersionPreLabel")
```
Would it make sense to change this property name to `PSSemanticVersionPrereleaseLabel`?

iSazonov replied:
Looks very long 😕

iSazonov replied:
Maybe ` PSSemVerPreReleaseLabel `?

daxian-dbw replied:
That looks good to me, especially `semver` is the type accelerator for `SemanticVersion`.

iSazonov replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -398,8 +406,39 @@ public SemanticVersion(string version)
+        /// If <paramref name="preLabel"/> don't match 'LabelUnitRegEx'.
+        /// If <paramref name="buildLabel"/> don't match 'LabelUnitRegEx'.
+        /// </exception>
+        public SemanticVersion(int major, int minor, int patch, string preLabel, string buildLabel)
```
Shall we rename `preLabel` to `prereleaseLabel` too? It's the parameter of a public method.

iSazonov replied:
Renamed.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -384,12 +386,18 @@ IEnumerator IEnumerable.GetEnumerator()
+        private const string VersionSansRegEx = @"^(?<major>\d+)(\.(?<minor>\d+))?(\.(?<patch>\d+))?$";
+        private const string LabelRegEx = @"^((?<preLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?(\+(?<buildLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?$";
+        private const string LabelUnitRegEx = @"^[0-9A-Za-z][0-9A-Za-z\-\.]*$";
+        private const string PreLabelPropertyName = "PSSemanticVersionPreLabel";
```
Same here. We probably should use `PSSemanticVersionPrereleaseLabel`, since it's for a public property.

iSazonov replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -571,77 +646,154 @@ public static bool TryParse(string version, out SemanticVersion result)
+        /// </summary>
+        public static int Compare(SemanticVersion versionA, SemanticVersion versionB)
+        {
+            if (versionA != null) {
```
Minor comment: it's better to follow the same open curly brace style in this file.

iSazonov replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -571,77 +646,154 @@ public static bool TryParse(string version, out SemanticVersion result)
+                if (versionB != null)
+                {
+                    return versionA.CompareTo(versionB);
+                }
```
The instance method `CompareTo` handles null value in it, so this `versionB != null` check is not needed. It can just be:
```
if (versionA != null)
{
    return versionA.CompareTo(versionB);
}

if (versionB != null)
{
    return -1;
}

return 0;
```

iSazonov replied:
Fixed.

daxian-dbw commented:
I still have a concern about the `GetHashCode()` method. See my new comment at https://github.com/PowerShell/PowerShell/pull/5037#discussion_r149441925

daxian-dbw commented:
@SteveL-MSFT This PR changes the public APIs, including:
- Add a public constructions `public SemanticVersion(int major, int minor, int patch, string preLabel, string buildLabel)`
- Remove public property `Label` and add two public properties `PreReleaseLabel` and `BuildLabel`.
- Add a public method `public static int Compare(SemanticVersion versionA, SemanticVersion versionB)`

Do you think it needs the committee review?

daxian-dbw commented on src/System.Management.Automation/engine/Types_Ps1Xml.cs:
```diff
@@ -1982,6 +1982,22 @@ public static IEnumerable<TypeData> Get()
+            td252.Members.Add("ToString",
+                new ScriptMethodData(@"ToString", GetScriptBlock(@"
+          $suffix = """"
+          if (![String]::IsNullOrEmpty($this.PSSemanticVersionPreLabel))
```
This needs to be changed if the property name is changed.

SteveL-MSFT commented:
@daxian-dbw the general guidance is that for small api changes, the code review process is sufficient.  However, any maintainer can request Committee review if needed or if the contributors for the code review can't come to an agreement.

iSazonov commented:
SemanticVersion API is new in PowerShell Core so until RTM we can change it free if we need.

daxian-dbw commented:
I'm fine with the API changes in this PR.

daxian-dbw commented on .vscode/tasks.json:
```diff
@@ -1,6 +1,6 @@
 {
     "version": "0.1.0",
-    "command": "pwsh",
+    "command": "powershell",
```
Please revert this change.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -530,10 +530,10 @@ public SemanticVersion(Version version)
             var psobj = new PSObject(version);
-            var preLabelNote = psobj.Properties[PreLabelPropertyName];
-            if (preLabelNote != null)
+            var PreReleaseLabelNote = psobj.Properties[PreLabelPropertyName];
```
A local variable name should start with a lower-case letter.
In fact, I think you can continue to use `preLabel` for any non-public members.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -387,10 +387,10 @@ IEnumerator IEnumerable.GetEnumerator()
     {
         private const string VersionSansRegEx = @"^(?<major>\d+)(\.(?<minor>\d+))?(\.(?<patch>\d+))?$";
-        private const string LabelRegEx = @"^((?<preLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?(\+(?<buildLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?$";
+        private const string LabelRegEx = @"^((?<PreReleaseLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?(\+(?<buildLabel>[0-9A-Za-z][0-9A-Za-z\-\.]*))?$";
```
Same here, I think it's OK to use `preLabel` here.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -656,7 +656,7 @@ private static bool TryParseVersion(string version, ref VersionResult result)
             var minor=0;
             var patch=0;
-            string preLabel = null;
+            string PreReleaseLabel = null;
```
Same here. Local variable should start with lower case letter.

daxian-dbw commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -917,7 +914,7 @@ public override int GetHashCode()
         }
 
-        private static int ComparePreLabel(string preLabel1, string preLabel2)
+        private static int ComparePreLabel(string PreReleaseLabel1, string PreReleaseLabel2)
```
Same here. parameter names.

daxian-dbw reviewed (commented):
It looks to me you just simply replaced `preLabel` with `PreReleaseLabel` and didn't care about the local variables and parameter names. Local variable names and parameter names should start with a lower case letter. I will fix it for you.

daxian-dbw committed:
[Feature] Address comment and fix immature replacement

daxian-dbw commented:
The test failure in AppVeyor is a known issue and will be fixed by #5379 
Travis CI passed, but it's not updated here.

iSazonov commented:
😕  Sorry, I forgot to set VS Code case-sensitive flag to search and replace.

daxian-dbw commented:
The macOS CI was shown as passed yesterday after clicking the Travis CI "Details" link. I have no idea why it's shown as canceled today. Now we seem to have a problem with macOS CI -- it takes forever to start.

Given that I know the Travis CI previously passed, I will merge this PR.
